### PR TITLE
Create package manager for registry

### PR DESCRIPTION
**Pull request by bryanl**
Creates package manager interface for registry. This is transition work for creating a better API around registries and packages. 

* Moves `pkg/Package` to an interface
* Creates `registry/PackageManager` interface

Signed-off-by: bryanl bryanliles@gmail.com

### CODE REVIEW

File: pkg/app/schema.go [hunk 1]
```diff
@@ -299,6 +299,7 @@ type EnvironmentDestinationSpec struct {
 type LibraryRefSpec struct {
 	Name       string          `json:"name"`
 	Registry   string          `json:"registry"`
+	Version    string          `json:"version"`
 	GitVersion *GitVersionSpec `json:"gitVersion,omitempty"`
```
Comment (shomron) [inline pkg/app/schema.go]: We can work on  removing `GitVersion` in a separate commit.

File: pkg/pkg/helm.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+		if err != nil {
+			return "", errors.Wrapf(err, "finding latest %s chart release", name)
+		}
+	}
+
+	chartConfigPath := filepath.Join(a.VendorPath(), registryName, name, "helm", version, name, "Chart.yaml")
+	return chartConfigPath, nil
+}
+
+func latestChartRelease(a app.App, name, registryName, version string) (string, error) {
```
Comment (shomron) [inline pkg/pkg/helm.go]: - [ ] `version` is unused.

File: pkg/pkg/helm.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+	return &Helm{
+		a:            a,
+		name:         name,
+		registryName: registryName,
+		config:       cc,
+	}, nil
+}
+
+func chartConfigPath(a app.App, name, registryName, version string) (string, error) {
```
Comment (shomron) [inline pkg/pkg/helm.go]: Please document that `version` is optional.

File: pkg/pkg/helm.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+			if err != nil {
+				return "", err
+			}
+
+			versions = append(versions, v)
+		}
+	}
+
+	semver.Sort(versions)
+	return versions[0].String(), nil
```
Comment (shomron) [inline pkg/pkg/helm.go]: - [ ] Need to check for empty `versions`

File: pkg/pkg/helm.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+// Description returns the description for the Helm chart. The description
+// is retrieved from the chart's Chart.yaml file.
+func (h *Helm) Description() string {
+	return h.config.Description
+}
+
+// IsInstalled returns true if the package is installed. a package is installed if it
+// has a libraries entry in app.yaml.
```
Comment (shomron) [inline pkg/pkg/helm.go]: Should we corroborate that with the contents of `vendor`? I've already confused the tool by deleting `vendor` and not updating `app.yaml` during testing.

File: pkg/pkg/helm.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+func (h *Helm) IsInstalled() (bool, error) {
```
Comment (shomron) [inline pkg/pkg/helm.go]: Should we share this implementation with other packages types?

File: pkg/pkg/helm.go [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+	_, isInstalled := libs[h.Name()]
+	return isInstalled, nil
+}
+
+func (h *Helm) prototypeName() string {
+	return fmt.Sprintf("io.ksonnet.pkg.%s-%s", h.registryName, h.name)
+}
+
+// Prototypes returns prototypes for this package. Currently, it returns a single prototype.
```
Comment (shomron) [inline pkg/pkg/helm.go]: Do we think that assumption will change for helm packages? (1:1)

File: pkg/pkg/local.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+	name         string
+	registryName string
+	config       *parts.Spec
+}
+
+var _ Package = (*Local)(nil)
+
+// NewLocal creates an instance of Local.
+func NewLocal(a app.App, name, registryName string) (*Local, error) {
+	partsPath := filepath.Join(a.VendorPath(), registryName, name, "parts.yaml")
```
Comment (shomron) [inline pkg/pkg/local.go]: Is ownership of the cache going to transition from `Registry` to `PackageManager`? Given the work I'm currently doing, the less components touching the cache directly, the better.

Comment (shomron) [reply pkg/pkg/local.go]: Wherever it lands, I'd like to see an interface centered around `io.Reader` / `io.Writer`, and less dependence on file path manipulation and `afero` dispersed across the code base.

File: pkg/pkg/local.go [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+// Description returns the description for the package. The description
+// is ready from the package's parts.yaml.
+func (l *Local) Description() string {
+	return l.config.Description
+}
+
+// IsInstalled returns true if the package is installed. a package is installed if it
+// has a libraries entry in app.yaml.
+func (l *Local) IsInstalled() (bool, error) {
+	// TODO: this code is duplicated in helm.go
```
Comment (shomron) [inline pkg/pkg/local.go]: Yep, noted above.

File: pkg/pkg/local.go [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		return prototypes, nil
+	}
+
+	err = afero.Walk(l.a.Fs(), protoPath, func(path string, fi os.FileInfo, err error) error {
+		if fi.IsDir() || filepath.Ext(path) != ".jsonnet" {
```
Comment (shomron) [inline pkg/pkg/local.go]: Should we support `.libsonnet`? Are these files meant to be imported?

File: pkg/registry/package_manager.go [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	libraryConfigs, err := m.app.Libraries()
+	if err != nil {
+		return nil, errors.Wrap(err, "reading libraries defined in the configuration")
+	}
+
+	libraryConfig, ok := libraryConfigs[d.Part]
```
[the remark this answers is not in the document]
Comment (bryanl) [reply pkg/registry/package_manager.go]: Yes. This is a bad thing waiting to happen.

File: pkg/pkg/helm.go (hunk 7 above)
Comment (bryanl) [reply pkg/pkg/helm.go]: I'm not sure, but don't want to code myself in a corner.

File: pkg/registry/package_manager.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (shomron) [reply pkg/registry/package_manager.go]: Opened #617

File: pkg/pkg/helm.go (hunk 7 above)
Comment (shomron) [reply pkg/pkg/helm.go]: 👍

File: pkg/pkg/local.go (hunk 10 above)
Comment (bryanl) [reply pkg/pkg/local.go]: prototypes have a `.jsonnet` extension by convention.

**Comment by bryanl** [comment]
All updates have been made

File: pkg/pkg/local.go [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+// Local is a package based on vendored contents.
+type Local struct {
+	a              app.App
+	name           string
+	registryName   string
+	config         *parts.Spec
+	installChecker InstallChecker
+}
+
+var _ Package = (*Local)(nil)
```
Comment (underrun) [inline pkg/pkg/local.go]: why do this? - looks like Google tells me that you get to force the compiler to make sure something is able to be cast as something else. but i don't understand when/why that is necessary that isn't covered by just using things normally or specifically implementing tests.

Comment (bryanl) [reply pkg/pkg/local.go]: It allows you to ensure a type implements a struct while editing it.

**Pull request merged**